Design proposal for custom Xcode scheme support (#592)
buildbuddy-io/rules_xcodeproj: doc/design/custom_xcode_schemes.md

  * [Specifying a Launch Target](#specifying-a-launch-target)
  * [Specifying Launch Arguments and Environment Variables](#specifying-launch-arguments-and-environment-variables)
* [Introduction of Scheme Autogeneration Mode](#introduction-of-scheme-autogeneration-mode)
* [Implementation Changes](#implementation-changes)
  * [xcode\_schemes Module](#xcode_schemes-module)
  * [Changes to \_xcodeproj Rule](#changes-to-_xcodeproj-rule)
    * [scheme\_autogeneration\_mode Attribute](#scheme_autogeneration_mode-attribute)
    * [schemes Attribute](#schemes-attribute)
  * [Changes to xcodeproj Macro](#changes-to-xcodeproj-macro)
    * [Collecting Targets from Schemes](#collecting-targets-from-schemes)
* [Configuration and Target Selection in Schemes](#configuration-and-target-selection-in-schemes)
* [Outstanding Questions](#outstanding-questions)

## Automatic Scheme Generation

As of this writing, the `rules_xcodeproj` ruleset generates an Xcode scheme for every buildable
target provided to the `xcodeproj` rule. This allows a client to quickly define an `xcodeproj`
target and generate an Xcode project.

Let's start with an example.

```python
# Assumptions

> #   //Sources/Foo:Foo - swift_library
> #   //Sources/FooTests:FooTests = ios_unit_test
> 
> xcodeproj(
>     name = "generate_xcodeproj",
>     project_name = "Command Line",
>     tags = ["manual"],
>     targets = [
>         "//Sources/Foo",

Reviewer: Unless we want to get into the intricacies of how this will actually create _3 targets_ and _3 schemes_ let's not include the library in `targets`.

```suggestion
#   //Sources/FooTests:FooTests = ios_unit_test
#   //Sources/Foo:Foo - swift_library, a dependency of //Sources/FooTests:FooTests

xcodeproj(
    name = "generate_xcodeproj",
    project_name = "Command Line",
    tags = ["manual"],
    targets = [
```

`xcodeproj` macro.

### Specifying a Launch Target

Let's continue our example. We will add an `ios_application` and a `ios_ui_test` to the mix.

```python
# Assumptions
#   //Sources/Foo:Foo - swift_library
#   //Sources/FooTests:FooTests = ios_unit_test
#   //Sources/App = ios_application
#   //Sources/AppUITests = ios_ui_test

load(
    "@com_github_buildbuddy_io_rules_xcodeproj//xcodeproj:xcodeproj.bzl",
    "xcode_schemes",
    "xcodeproj",
)

_SCHEMES = [
    xcode_schemes.scheme(
        name = "Foo Module",
        build_action = xcode_schemes.build_action(["//Sources/Foo"]),
        test_action = xcode_schemes.test_action(["//Tests/FooTests"]),
    ),
    xcode_schemes.scheme(
        name = "My Application",
        test_action = xcode_schemes.test_action([
            "//Tests/AppUITests",
            "//Tests/FooTests",
        ]),
        launch_action = xcode_schemes.launch_action("//Sources/App"),
    ),
]

xcodeproj(
    name = "generate_xcodeproj",
    project_name = "Command Line",
    tags = ["manual"],
    schemes = _SCHEMES,
)
```

The above example adds a second scheme called `My Application`. It is configured with two test targets

> and a single launch target. All three targets are implicitly included in the build list.

Reviewer: >All three targets are implicitly included in the build list.

Is `//Tests/AppUITests` set to build for `Run` or only `Test` (default behavior would only be `Test`). And are we able to force it to build during `Run` by listing it explicitly in `build_action`?

Author: Good question. For reference, here is the [BuildFor enum](https://github.com/tuist/XcodeProj/blob/3a93b47a34860a4d7dbcd9cc0ae8e9543c179c61/Sources/XcodeProj/Scheme/XCScheme%2BBuildAction.swift#L8-L13).

I was thinking that we would just set the `buildFor` to `default`.  For autogenerated schemes, we set the following list:
```swift
                buildFor: [
                    .running,
                    .testing,
                    .profiling,
                    .archiving,
                    .analyzing
                ]
```

I think that we can be as smart of dumb about this as we want. If you think that it is important to be selective, we can add it. I don't have a good sense for whether folks customize this for their real projects.

Reviewer: I believe Xcode defaults to having test targets `buildFor: [.testing]`.

I say anything we do this:
- Anything `build_action` gets `.testing` in `buildFor`
- Anything in `launch_action` or `build_action` gets `[.running, .profiling, .archiving, .analyzing]` in `buildFor`
- Thus being in both gets all of them

Author: Did you mean `test_action` for the first bullet?

Author: The rest sounds reasonable to me.

Author: I added a section enumerating the build for logic.

### Specifying Launch Arguments and Environment Variables

> There are times when it is desirabe to customize the launch action with arguments and environment

Reviewer 2: Another launch configuration option that I use often is to specify a custom working directory. Is this something we could support configuring here?

Reviewer: I think so, yes.

Author: Sure. Do you anticipate hardcoding an absolute path in the BUILD file or would you be using `$(location)`?

Reviewer 2: What would `$(location)` resolve to? When I'm building CLIs I often want to run them in the project's root directory or an absolute directory.

Author: Digging into `tuist/XcodeProj`, I see [useCustomWorkingDirectory](https://github.com/tuist/XcodeProj/blob/3a93b47a34860a4d7dbcd9cc0ae8e9543c179c61/Sources/XcodeProj/Scheme/XCScheme%2BLaunchAction.swift#L45) that specifies that a custom working directory can be used. However, I don't see [customWorkingDirectory in the data model](https://github.com/tuist/XcodeProj/search?q=customWorkingDirectory). I confirmed that is the value that is written when you specify a custom working directory in Xcode. 

It looks like we will need to put up a PR to tuist to add it.

Author: I have updated the doc with `working_directory` support for `launch_action`.

> There are times when it is desirabe to customize the launch action with arguments and environment

Reviewer: ```suggestion
There are times when it is desirable to customize the launch action with arguments and environment
```

    "xcodeproj",
)

_SCHEMES = [
    xcode_schemes.scheme(
        name = "Foo Module",
        build_action = xcode_schemes.build_action(["//Sources/Foo"]),
        test_action = xcode_schemes.test_action(["//Tests/FooTests"]),
    ),
    xcode_schemes.scheme(
        name = "My Application",
        test_action = xcode_schemes.test_action([
            "//Tests/AppUITests",
            "//Tests/FooTests",
        ]),
        launch_action = xcode_schemes.launch_action(
            target = "//Sources/App",
            args = [
                "--my_awesome_flag",
                "path/to/a/file.txt",
            ],
            env = {
                "RELEASE_THE_KRAKEN": "true",
            },
        ),
    ),
]

xcodeproj(
    name = "generate_xcodeproj",
    project_name = "Command Line",
    tags = ["manual"],
    schemes = _SCHEMES,
)
```

This example is the same as the previous one except arguments and environment variables are
defined for the launch action.

## Introduction of Scheme Autogeneration Mode

With this proposal, it is now possible to control how Xcode scheme autogeneration works using
the `scheme_autogeneration_mode` attribute on `_xcodeproj`. 

__Values__

> - `auto`: If no custom schemes are provided, an Xcode scheme will be created for every buildable

Reviewer 2: If there are no custom schemes provided and the mode is `auto`, why wouldn't schemes be created just for the list of `targets` explicitly specified? To get schemes for all buildable transitive deps you'd need to specify `all`.

Author: Sorry. My wording is imprecise. The intention is for `auto` to be work like `all` when no custom schemes exist. I was trying to preserve the current behavior.

Reviewer 2: That part was actually clear to me, I'm suggesting an alternative which would allow reducing the scheme pollution by default when using `targets` instead of `schemes` while still having a way to generate schemes for all transitive deps using `all`.

Reviewer: I could see that being another option, where it's `top_level_only`.

Reviewer 2: At Lyft, our generated Xcode project can be configured to go up and/or down N levels following the dependency graph for either transitive or reverse dependencies.

Reviewer: But that's Focused Projects, not scheme control, right? Focused Projects will only have schemes for targets that are in-focus, so the behavior would be the same then.

So if you had Focused Projects, would you use need this different generation mode?

Reviewer 2: I was under the impression that a "focused project" meant that there would only be schemes only for a subset of the build graph. Are those not the same thing?

Reviewer: Focused Project means that there will only be targets and their associated files in Xcode. (It's not just schemes that slow down Xcode, the targets themselves do as well.) Schemes can only be for targets that are in Xcode, so if you have a Focused Project it will have fewer schemes. But you can have a non-Focused Project (maybe you want all the targets and files) and still want fewer schemes, which is where I thought `top_level_only` might be useful.

Reviewer 2: ok, so if I'm understanding correctly, limiting the created schemes to a subset of the build graph doesn't mean the project is focused.

Only limiting the source files or targets to a subset of the build graph is referred to as a focused project.

Reviewer: Correct. We will need a glossary soon 😅.

Reviewer 3: I really like the addition of `top_level_only`. In our project we have schemes for the top-level buildables (such as apps and extensions mostly). These would cover the majority of the use-cases. Via the separate API it would then be possible to specify additional schemes for other library and test targets.

Author: So, to close this out, do we want the `top_level_only` value? Will you use it day one?

Reviewer: We want it, and it seems @BalestraPatrick would use it regardless of Focused Projects support.

Reviewer 2: I'd use `top_level_only` even without custom schemes if it's available before because I do see a perf hit in the scheme selector that is improved when trimming the schemes.

Author: I will add it.

  target. If custom schemes are provided, no autogenerated schemes will be created.

> - `none`: No custom schemes are generated.

Reviewer: ```suggestion
- `none`: No schemes are automatically generated.
```

> - `all`: Generate a scheme for every buildable target even if custom schemes are provided.

Reviewer: ```suggestion
- `all`: A scheme is generated for every buildable target even if custom schemes are provided.
```

The default value for `scheme_autogeneration_mode` is `auto`.

## Implementation Changes

### `xcode_schemes` Module

```python

> 

Reviewer: ```suggestion
```

def _scheme(
    name, 
    as_json = True, 
    build_action = None, 
    test_action = None, 
    launch_action = None):
    """Returns a `struct` or JSON `string` representing an Xcode scheme.

    Args:
        name: The user-visible name for the scheme as a `string`.
        as_json: Optional. A `bool` indicating whether the resulting struct 

>             should be returned as JSON.

Reviewer: What's the use case for this?

Author: This provides some syntactic sugar for the client. In the implementation, I was planning for all of the functions in `xcode_schemes` to return `struct` values. However, when the client uses `xcode_schemes.scheme`, the attribute on `_xcodeproj` rule will accept a `list` of JSON `string` values. I didn't the client to write something ugly like `xcode_schemes.scheme_json`. So, I added the `as_json` parameter. The client won't touch it, but our implementation and tests would.

As I was writing the above, it occurred to me that we can go at this a slightly different way. Today, we wrap calls to `_xcodeproj` in the macro `xcodeproj`. We could have `xcode_schemes.scheme` return a `struct` like the other functions. The `xcodeproj` macro can then serialize them to JSON before passing them to `_xcodeproj`. This approach has another nice feature. It skips a serialization/deserialization roundtrip. 

What do you think?

Reviewer: >The xcodeproj macro can then serialize them to JSON before passing them to _xcodeproj

That is what I was thinking as well. Let's go that route.

### Changes to `xcodeproj` Macro 

The `xcodeproj` macro will gain two parameters: `targets` and `schemes`.

```python
def xcodeproj(*, 
    name, 
    xcodeproj_rule = _xcodeproj, 
    targets = [], 
    schemes = None, 
    **kwargs):
    """Creates an .xcodeproj file in the workspace when run.

    Args:
        name: The name of the target.
        xcodeproj_rule: The actual `xcodeproj` rule. This is overridden during
            fixture testing. You shouldn't need to set it yourself.
        targets: Optional. A `list` of targets to be included in the Xcode 
            project.
        schemes: Optional. A `list` of JSON `string` values as returned by
            `xcode_schemes.scheme`.
        **kwargs: Additional arguments to pass to `xcodeproj_rule`.
    """
    pass
```

In addition to what the `xcodeproj` macro does today, if a value is provided for the `schemes`
parameter, the macro will collect all of the targets listed in the test and launch actions (not the
build action) and add them to the overall targets list. We only collect targets from the test action
and launch action because the target list for `xcodeproj` should only contain top-level or

> leaf-nodes (i.e., not depeneded upon by other targets in the Xcode project).

Reviewer: > We only collect targets from the test action and launch action because the target list for `xcodeproj` should only contain top-level or leaf-nodes (i.e., not depeneded upon by other targets in the Xcode project).

A library _could_ be a top-level target though, and it might not have a test. I think it's fine to force them to re-list that target in `targets`, but it's probably worth mentioning.

Also, we need to make sure to allow duplications (i.e. someone listing out all top level targets in `targets` and specifying custom schemes for each).

    # list.
    targets = [
        "//Sources/App",
        "//Tests/FooTests",
        "//Tests/AppUITests",
    ],
)
```

Note that the `targets` list is a deduplicated, sorted list of the targets that were provided from
the schemes and directly in the `targets` parameter.

The JSON representation of `Foo Module` will look like the following:

```json
{
  "name": "Foo Module",
  "build_action": {
    "targets": [
      "//Sources/Foo",
      "//Tests/FooTests"
    ],
  },
  "test_action": {
    "targets": [
      "//Tests/FooTests"
    ]
  }
}
```

Note that the build action targets list contains the targets from the build action declaration and
the test action declaration.

The JSON representation of `My Application` will look like the following:

```json
{
  "name": "My Application",
  "build_action": {
    "targets": [
      "//Sources/App",
      "//Tests/FooTests"
      "//Tests/AppUITests"
    ],
  },
  "test_action": {
    "targets": [
      "//Tests/FooTests",
      "//Tests/AppUITests"
    ]
  },
  "launch_action": {
    "target": "//Sources/App"
  }
}
```

> TODO: Is selecting targets from the test actions and launch actions sufficient for only incuding

> > top-level targets in the overall targets list?

Reviewer: What do you mean by this?

Author: In short, are you OK with the technique of populating the the `targets` list for `_xcodeproj` rule using the targets listed under `test_action` and `launch_action`?

Reviewer: Yes.

specified by their Bazel target only. The logic inside `rules_xcodeproj` will select the correct
configuration variant based upon the leaf nodes that are included in the scheme.

## Build for Configuration Logic

> Xcode schemes support configuraiton that dictates when a target is built. In the `tuist/XcodeProj`

Reviewer: ```suggestion
Xcode schemes support configuration that dictates when a target is built. In the `tuist/XcodeProj`
```

data model, this is modeled under
[`XCScheme.BuildAction.Entry.BuildFor`](https://github.com/tuist/XcodeProj/blob/3a93b47a34860a4d7dbcd9cc0ae8e9543c179c61/Sources/XcodeProj/Scheme/XCScheme%2BBuildAction.swift#L8-L13).

> The following will be the logic used to set the `BuildFor` for targets listed in custom schemes.

Reviewer: ```suggestion

The following will be the logic used to set the `BuildFor` for targets listed in custom schemes:
```
